Sources/_StringProcessing/ByteCodeGen.swift

```diff
@@ -292,9 +299,41 @@ extension Compiler.ByteCodeGen {
     }
   }
 
+  mutating func resolveReferences() throws {
+    for (id, uses) in unresolvedReferences {
+      guard let offset = captureOffsets[id] else {
+        throw RegexCompilationError.uncapturedReference
+      }
+      for use in uses {
+        builder.instructions[use.rawValue] =
+          Instruction(.backreference, .init(capture: .init(offset)))
+      }
+    }
+  }
+
+  mutating func emitUnresolvedReference(id: Reference.ID) {
+    builder.buildBackreference(.init(0))
+    unresolvedReferences[id, default: []].insert(builder.lastInstructionAddress)
+  }
+
+  mutating func registerCapture(id: Reference.ID?) {
+    if let id = id {
+      let preexistingValue = captureOffsets.updateValue(
+        captureCount, forKey: id)
+      assert(preexistingValue == nil)
+    }
+    captureCount += 1
+  }
+
   mutating func emitGroup(
-    _ kind: AST.Group.Kind, _ child: DSLTree.Node
+    _ kind: AST.Group.Kind,
+    _ child: DSLTree.Node,
+    _ referenceID: Reference.ID?
   ) throws -> CaptureRegister? {
+    guard kind.isCapturing || referenceID == nil else {
+      throw Unreachable("Reference ID shouldn't exist for non-capturing groups")
+    }
```

Comment:
I'm also very strongly considering splitting off the capturing concept into its own DSLTree node. But that's a future PR.

Reply:
Yes, I agree. The current construction is a bit weird. Perhaps something like `case capture(DSLTree.Node, CaptureTransform?, ReferenceID?)`